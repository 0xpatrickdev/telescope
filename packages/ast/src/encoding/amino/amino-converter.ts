import * as t from '@babel/types';
import { ProtoType } from '../types';
import { ProtoRoot, ProtoAny, AminoException, AminoExceptions, DEFAULT_AMINO_EXCEPTIONS } from '../../types';
import { toAminoJsonMethod } from './to-amino-json';
import { fromAminoJsonMethod } from './from-amino-json';
import { kebab } from 'case';
import { getTypeUrl, typeUrlToAmino } from './utils';

export interface AminoOptions {
    aminoCasingFn: Function,
    exceptions?: AminoExceptions
}


// MARKED AS NOT DRY
interface Context {
    enums: any[];
    types: any[];
}

export interface AminoConverterItemParams {
    root: ProtoRoot,
    context: Context,
    proto: ProtoType,
    options: AminoOptions,
}
export const makeAminoConverterItem = (
    {
        root,
        context,
        proto,
        options
    }: AminoConverterItemParams
) => {

    const typeUrl = getTypeUrl(root, proto);

    return t.objectProperty(
        t.stringLiteral(typeUrl),
        t.objectExpression(
            [
                t.objectProperty(t.identifier('aminoType'), t.stringLiteral(typeUrlToAmino(typeUrl, options.exceptions))),
                t.objectProperty(
                    t.identifier('toAmino'),
                    toAminoJsonMethod(
                        context,
                        proto,
                        options
                    )
                ),
                t.objectProperty(
                    t.identifier('fromAmino'),
                    fromAminoJsonMethod(
                        proto,
                        options
                    )
                )
            ]
        )
    );
};


interface AminoConverterParams {
    name: string,
    root: ProtoRoot,
    context: Context,
    protos: ProtoAny[],
    options: AminoOptions,
    exceptions?: AminoExceptions
}
export const aminoConverter = (
    {
        name,
        root,
        context,
        protos,
        options,
        exceptions
    }: AminoConverterParams) => {

    const items = protos.map(proto => {
        return makeAminoConverterItem({
            context,
            root,
            proto,
            options,
            exceptions
        })
    })

    return t.exportNamedDeclaration(t.variableDeclaration('const', [
        t.variableDeclarator(t.identifier(name),
            t.objectExpression(
                items
            ))
    ]));
};

