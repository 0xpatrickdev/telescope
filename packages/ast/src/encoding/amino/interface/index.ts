import * as t from '@babel/types';
import { getTSTypeFromProtoType, ProtoField, ProtoType } from '../../proto/types';
import { AminoOptions, AminoParseContext } from '../types';
import { getTypeUrl, protoFieldsToArray, typeUrlToAmino } from '../utils';
import { aminoInterface } from './utils';

export interface RenderAminoField {
    context: AminoParseContext;
    field: ProtoField;
    options: AminoOptions;
};

export const renderAminoField = ({
    context,
    field,
    options
}: RenderAminoField) => {

    if (field.rule === 'repeated') {
        switch (field.parsedType.type) {
            case 'Type':
                return aminoInterface.typeArray({
                    context,
                    field,
                    options
                });
            case 'Enum':
                return aminoInterface.enumArray(
                    field,
                    options
                );
            default:
                return aminoInterface.array(field, options);
        }
    }

    switch (field.parsedType.type) {
        case 'Type':
            return aminoInterface.type({
                context,
                field,
                options
            });

        case 'Enum':
            return aminoInterface.enum(
                field,
                options
            );
    }

    // scalar types...
    switch (field.type) {
        case 'string':
            return aminoInterface.defaultType(field, options, getTSTypeFromProtoType(field.type));
        case 'int64':
        case 'uint64':
            return aminoInterface.long(field, options);
        case 'double':
        case 'int64':
        case 'bool':
        case 'bytes':
        case 'Timestamp':
        case 'google.protobuf.Timestamp':
            return aminoInterface.defaultType(field, options, getTSTypeFromProtoType(field.type));

        // TODO check can we just
        // make pieces optional and avoid hard-coding this type?
        case 'ibc.core.client.v1.Height':
        case 'Height':
            return aminoInterface.height(field, options);

        case 'Duration':
        case 'google.protobuf.Duration':
            return aminoInterface.defaultType(field, options, t.tsStringKeyword());

        default:
            return aminoInterface.defaultType(field, options, getTSTypeFromProtoType(field.type));
    }
};

export interface MakeAminoTypeInterface {
    context: AminoParseContext;
    proto: ProtoType;
    options: AminoOptions;
};

export const makeAminoTypeInterface = ({
    context,
    proto,
    options
}: MakeAminoTypeInterface) => {


    const TypeName = proto.name;
    const typeUrl = getTypeUrl(context.ref.proto, proto);
    const aminoType = typeUrlToAmino(typeUrl, options.exceptions);

    return t.exportNamedDeclaration(
        t.tsInterfaceDeclaration(
            t.identifier('Amino' + TypeName),
            null,
            [t.tsExpressionWithTypeArguments(t.identifier('AminoMsg'))],
            t.tSInterfaceBody([
                t.tSPropertySignature(t.identifier('type'), t.tsTypeAnnotation(
                    t.tSLiteralType(t.stringLiteral(aminoType))
                )),
                t.tSPropertySignature(t.identifier('value'), t.tsTypeAnnotation(t.tsTypeLiteral(
                    protoFieldsToArray(proto).map((field) => {
                        return renderAminoField({
                            context,
                            field,
                            options
                        });
                    })
                )))
            ])
        )
    )

}