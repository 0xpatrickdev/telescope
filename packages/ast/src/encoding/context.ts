import { ProtoStore, ProtoRef } from '@osmonauts/proto-parser';

export interface ParseContext {
    utils: Record<string, boolean>;
    addUtil: Function;
}

export class GenericParseContext implements ParseContext {
    imports: Record<string, object> = {};
    utils: Record<string, boolean> = {};

    addUtil(util) {
        this.utils[util] = true;
    }
}
export class AminoParseContext implements ParseContext {
    store: ProtoStore;
    ref: ProtoRef;

    utils: Record<string, boolean> = {};

    constructor(
        ref: ProtoRef,
        store: ProtoStore
    ) {
        this.ref = ref;
        this.store = store;
    }

    addUtil(util) {
        this.utils[util] = true;
    }
}

export class ProtoParseContext implements ParseContext {
    imports: Record<string, object> = {};
    utils: Record<string, boolean> = {};

    addUtil(util) {
        this.utils[util] = true;
    }
}

