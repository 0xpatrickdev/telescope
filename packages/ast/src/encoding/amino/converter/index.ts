import * as t from '@babel/types';
import { ProtoType } from '../../proto/types';
import { ProtoRoot } from '@osmonauts/proto-parser';
import { AminoOptions } from '../types';
import { toAminoJsonMethod } from '../to-amino-json';
import { fromAminoJsonMethod } from '../from-amino-json';
import { getTypeUrl, typeUrlToAmino } from '../utils';
import { ParseContext } from '../../context';

interface AminoConverterItemParams {
    root: ProtoRoot,
    context: ParseContext,
    proto: ProtoType,
    options: AminoOptions
}
export const makeAminoConverterItem = (
    {
        root,
        context,
        proto,
        options
    }: AminoConverterItemParams
) => {

    const typeUrl = getTypeUrl(root, proto);

    return t.objectProperty(
        t.stringLiteral(typeUrl),
        t.objectExpression(
            [
                t.objectProperty(t.identifier('aminoType'), t.stringLiteral(typeUrlToAmino(typeUrl, options.exceptions))),
                t.objectProperty(
                    t.identifier('toAmino'),
                    toAminoJsonMethod({
                        context,
                        proto,
                        options
                    })
                ),
                t.objectProperty(
                    t.identifier('fromAmino'),
                    fromAminoJsonMethod({
                        context,
                        proto,
                        options
                    })
                )
            ]
        )
    );
};




interface AminoConverterParams {
    name: string,
    root: ProtoRoot,
    context: ParseContext,
    protos: ProtoType[],
    options: AminoOptions
}
export const aminoConverter = (
    {
        name,
        root,
        context,
        protos,
        options
    }: AminoConverterParams) => {

    const items = protos.map(proto => {
        return makeAminoConverterItem({
            context,
            root,
            proto,
            options
        })
    })

    return t.exportNamedDeclaration(t.variableDeclaration('const', [
        t.variableDeclarator(t.identifier(name),
            t.objectExpression(
                items
            ))
    ]));
};

