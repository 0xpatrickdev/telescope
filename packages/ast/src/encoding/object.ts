import * as t from '@babel/types';
import { ProtoType } from './proto/types';
import { fromPartialMethod } from './proto/from-partial';
import { decodeMethod } from './proto/decode';
import { encodeMethod } from './proto/encode';
import { fromJSONMethod } from './proto/from-json';
import { toJSONMethod } from './proto/to-json';
import { ProtoParseContext } from './context';

export const createObjectWithMethods = (
    context: ProtoParseContext,
    name: string,
    proto: ProtoType
) => {

    const methods = [
        encodeMethod(context, name, proto),
        decodeMethod(context, name, proto),
        fromJSONMethod(context, name, proto),
        toJSONMethod(context, name, proto),
        fromPartialMethod(context, name, proto),
    ];

    return t.exportNamedDeclaration(
        t.variableDeclaration('const',
            [
                t.variableDeclarator(
                    t.identifier(name),
                    t.objectExpression(
                        methods
                    )
                )
            ]
        )
    )
};