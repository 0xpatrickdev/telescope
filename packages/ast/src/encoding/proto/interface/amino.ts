import * as t from '@babel/types';
import { ProtoType, ProtoField, ProtoEnum } from '@osmonauts/types';
import { identifier, tsPropertySignature, makeCommentBlock } from '../../../utils';
import { ProtoParseContext } from '../../context';

import {
    getFieldOptionality,
    getOneOfs
} from '../types';

import {
    CreateProtoTypeOptions,
    SymbolNames,
    getFieldAminoTypeReference,
    getTSAminoType
} from '../../types';
import { getAminoTypeName } from '../../amino';

export const createAminoTypeOptionsDefaults: CreateProtoTypeOptions = {
    useOriginalCase: true,
    typeNameSuffix: 'Amino'
};

const getAminoField = (
    context: ProtoParseContext,
    field: ProtoField,
    options: CreateProtoTypeOptions = createAminoTypeOptionsDefaults
) => {
    let ast: any = null;

    ast = getFieldAminoTypeReference(context, field, options);

    if (field.rule === 'repeated') {
        ast = t.tsArrayType(ast);
    }

    if (field.keyType) {
        ast = t.tsUnionType([
            t.tsTypeLiteral([
                t.tsIndexSignature([
                    identifier('key',
                        t.tsTypeAnnotation(
                            getTSAminoType(context, field.keyType)
                        )
                    )
                ],
                    t.tsTypeAnnotation(ast)
                )
            ])
        ]);
    }

    return ast;
};

export const createAminoType = (
    context: ProtoParseContext,
    name: string,
    proto: ProtoType,
    options: CreateProtoTypeOptions = createAminoTypeOptionsDefaults
) => {
    const oneOfs = getOneOfs(proto);

    const MsgName = SymbolNames.Msg(name, options);

    // scalar amino types!
    let declaration: t.ExportNamedDeclaration | undefined;
    if (proto.type === 'Type') {
        switch (proto.name) {
            case 'Duration':
            case 'google.protobuf.Duration':
            case 'Timestamp':
            case 'google.protobuf.Timestamp':
                declaration = t.exportNamedDeclaration(
                    t.tsTypeAliasDeclaration(
                        t.identifier(MsgName),
                        null,
                        t.tsStringKeyword()
                    )
                );
                break;
            default:
        }
    }

    // declaration
    if (!declaration) {
        declaration = t.exportNamedDeclaration(t.tsInterfaceDeclaration(
            t.identifier(MsgName),
            null,
            [],
            t.tsInterfaceBody(
                Object.keys(proto.fields).reduce((m, fieldName) => {
                    const isOneOf = oneOfs.includes(fieldName);
                    const field = proto.fields[fieldName];

                    const orig = field.options?.['(telescope:orig)'] ?? fieldName;
                    let fieldNameWithCase = options.useOriginalCase ? orig : fieldName;

                    // should we actually just edit/add comments 
                    // to make this more "native" for any google.protobuf.Any?
                    // let's see...
                    if (
                        name === 'Any' &&
                        context.ref.proto.package === 'google.protobuf' &&
                        options.typeNameSuffix === 'Amino' &&
                        orig === 'type_url'
                    ) {
                        // type_url => type
                        fieldNameWithCase = 'type';
                    }

                    const propSig = tsPropertySignature(
                        t.identifier(fieldNameWithCase),
                        t.tsTypeAnnotation(
                            getAminoField(context, field, options)
                        ),
                        getFieldOptionality(context, field, isOneOf)
                    );

                    const comments = [];
                    if (field.comment) {
                        comments.push(
                            makeCommentBlock(field.comment)
                        );
                    }
                    if (field.options?.deprecated) {
                        comments.push(
                            makeCommentBlock('@deprecated')
                        );
                    }
                    if (comments.length) {
                        propSig.leadingComments = comments;
                    }

                    m.push(propSig)
                    return m;
                }, [])
            )
        ));
    }

    const comments = [];

    if (proto.comment) {
        comments.push(makeCommentBlock(proto.comment));
    }

    if (proto.options?.deprecated) {
        comments.push(makeCommentBlock('@deprecated'));
    }

    if (comments.length) {
        declaration.leadingComments = comments;
    }


    return declaration;
};

export const createAminoTypeType = (
    context: ProtoParseContext,
    name: string,
    proto: ProtoType,
    options: CreateProtoTypeOptions = createAminoTypeOptionsDefaults
) => {
    const MsgName = SymbolNames.Msg(name, options);
    const AminoMsgName = SymbolNames.AminoType(name, options);

    const aminoName = getAminoTypeName(context, context.ref.proto, proto);
    const typ = aminoName ? t.tsLiteralType(
        t.stringLiteral(aminoName)
    ) : t.tsTypeReference(
        t.identifier('string')
    );
    // scalar amino types!
    return t.exportNamedDeclaration(t.tsInterfaceDeclaration(
        t.identifier(AminoMsgName),
        null,
        [],
        t.tsInterfaceBody([
            tsPropertySignature(
                t.identifier('type'),
                t.tsTypeAnnotation(
                    typ
                ),
                false
            ),
            tsPropertySignature(
                t.identifier('value'),
                t.tsTypeAnnotation(
                    t.tsTypeReference(
                        t.identifier(MsgName)
                    )
                ),
                false
            )
        ])));
};



export const createEnumAminoType = (
    context: ProtoParseContext,
    name: string,
    proto: ProtoEnum
) => {
    return t.exportNamedDeclaration(
        t.variableDeclaration('const', [
            t.variableDeclarator(
                t.identifier(name + 'Amino'),
                t.identifier(name)
            )
        ])
    );
    // return createProtoEnum(
    //     context,
    //     name + 'Amino',
    //     proto
    // );
};
