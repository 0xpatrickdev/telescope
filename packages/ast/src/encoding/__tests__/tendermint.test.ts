import { getNestedProto } from '@osmonauts/proto-parser';
import { defaultTelescopeOptions } from '@osmonauts/types';
import { ProtoStore } from '@osmonauts/proto-parser'
import { expectCode } from '../../../test-utils/'
import { AminoParseContext, ProtoParseContext } from '../context';
import { createProtoType } from '..';
import { createObjectWithMethods } from '../object';
import { createAminoConverter } from '../amino/index';

const store = new ProtoStore([__dirname + '/../../../../../__fixtures__/chain1']);
store.traverseAll();

describe('ibc/lightclients/tendermint/v1/tendermint', () => {
    const ref = store.findProto('ibc/lightclients/tendermint/v1/tendermint.proto');
    const context = new ProtoParseContext(ref, store, defaultTelescopeOptions);
    const aminoCtx = new AminoParseContext(ref, store, defaultTelescopeOptions);
    describe('duration', () => {
        beforeAll(() => {
            context.options.useDuration = 'duration';
            aminoCtx.options.useDuration = 'duration';
        });
        it('interface', () => {
            expectCode(createProtoType(context, 'ClientState',
                getNestedProto(ref.traversed).ClientState
            ));
        });
        it('proto', () => {
            expectCode(createObjectWithMethods(context, 'ClientState',
                getNestedProto(ref.traversed).ClientState
            ));
        });
        // it('amino', () => {
        //     expectCode(createAminoConverter({
        //         context: aminoCtx,
        //         root: ref.proto,
        //         name: 'AminoConverter',
        //         protos: [getNestedProto(ref.traversed).ClientState]
        //     }));
        // });
    });
    describe('string', () => {
        beforeAll(() => {
            context.options.useDuration = 'string';
            aminoCtx.options.useDuration = 'string';
        });
        it('interface', () => {
            expectCode(createProtoType(context, 'ClientState',
                getNestedProto(ref.traversed).ClientState
            ));
        });
        it('proto', () => {
            expectCode(createObjectWithMethods(context, 'ClientState',
                getNestedProto(ref.traversed).ClientState
            ));
        });
        // it('amino', () => {
        //     expectCode(createAminoConverter({
        //         context: aminoCtx,
        //         root: ref.proto,
        //         name: 'AminoConverter',
        //         protos: [getNestedProto(ref.traversed).ClientState]
        //     }));
        // });
    })
});