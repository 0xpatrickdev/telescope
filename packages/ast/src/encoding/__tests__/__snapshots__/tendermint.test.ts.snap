// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ibc/lightclients/tendermint/v1/tendermint duration interface 1`] = `
"/**
 * ClientState from Tendermint tracks the current validator set, latest height,
 * and a possible frozen height.
 */
export interface ClientState {
  chainId: string;
  trustLevel: Fraction;

  /**
   * duration of the period since the LastestTimestamp during which the
   * submitted headers are valid for upgrade
   */
  trustingPeriod: string;

  /** duration of the staking unbonding period */
  unbondingPeriod: string;

  /** defines how much new (untrusted) header's Time can drift into the future. */
  maxClockDrift: string;

  /** Block height when the client was frozen due to a misbehaviour */
  frozenHeight: Height;

  /** Latest height the client was updated to */
  latestHeight: Height;

  /** Proof specifications used in verifying counterparty state */
  proofSpecs: ProofSpec[];

  /**
   * Path at which next upgraded client will be committed.
   * Each element corresponds to the key for a single CommitmentProof in the
   * chained proof. NOTE: ClientState must stored under
   * \`{upgradePath}/{upgradeHeight}/clientState\` ConsensusState must be stored
   * under \`{upgradepath}/{upgradeHeight}/consensusState\` For SDK chains using
   * the default upgrade module, upgrade_path should be []string{\\"upgrade\\",
   * \\"upgradedIBCState\\"}\`
   */
  upgradePath: string[];

  /**
   * This flag, when set to true, will allow governance to recover a client
   * which has expired
   */
  allowUpdateAfterExpiry: boolean;

  /**
   * This flag, when set to true, will allow governance to unfreeze a client
   * whose chain has experienced a misbehaviour event
   */
  allowUpdateAfterMisbehaviour: boolean;
}"
`;

exports[`ibc/lightclients/tendermint/v1/tendermint duration proto 1`] = `
"export const ClientState = {
  encode(message: ClientState, writer: _m0.Writer = _m0.Writer.create()): _m0.Writer {
    if (message.chainId !== \\"\\") {
      writer.uint32(10).string(message.chainId);
    }

    if (message.trustLevel !== undefined) {
      Fraction.encode(message.trustLevel, writer.uint32(18).fork()).ldelim();
    }

    if (message.trustingPeriod !== undefined) {
      Duration.encode(message.trustingPeriod, writer.uint32(26).fork()).ldelim();
    }

    if (message.unbondingPeriod !== undefined) {
      Duration.encode(message.unbondingPeriod, writer.uint32(34).fork()).ldelim();
    }

    if (message.maxClockDrift !== undefined) {
      Duration.encode(message.maxClockDrift, writer.uint32(42).fork()).ldelim();
    }

    if (message.frozenHeight !== undefined) {
      Height.encode(message.frozenHeight, writer.uint32(50).fork()).ldelim();
    }

    if (message.latestHeight !== undefined) {
      Height.encode(message.latestHeight, writer.uint32(58).fork()).ldelim();
    }

    for (const v of message.proofSpecs) {
      ProofSpec.encode(v!, writer.uint32(66).fork()).ldelim();
    }

    for (const v of message.upgradePath) {
      writer.uint32(74).string(v!);
    }

    if (message.allowUpdateAfterExpiry === true) {
      writer.uint32(80).bool(message.allowUpdateAfterExpiry);
    }

    if (message.allowUpdateAfterMisbehaviour === true) {
      writer.uint32(88).bool(message.allowUpdateAfterMisbehaviour);
    }

    return writer;
  },

  decode(input: _m0.Reader | Uint8Array, length?: number): ClientState {
    const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseClientState();

    while (reader.pos < end) {
      const tag = reader.uint32();

      switch (tag >>> 3) {
        case 1:
          message.chainId = reader.string();
          break;

        case 2:
          message.trustLevel = Fraction.decode(reader, reader.uint32());
          break;

        case 3:
          message.trustingPeriod = Duration.decode(reader, reader.uint32());
          break;

        case 4:
          message.unbondingPeriod = Duration.decode(reader, reader.uint32());
          break;

        case 5:
          message.maxClockDrift = Duration.decode(reader, reader.uint32());
          break;

        case 6:
          message.frozenHeight = Height.decode(reader, reader.uint32());
          break;

        case 7:
          message.latestHeight = Height.decode(reader, reader.uint32());
          break;

        case 8:
          message.proofSpecs.push(ProofSpec.decode(reader, reader.uint32()));
          break;

        case 9:
          message.upgradePath.push(reader.string());
          break;

        case 10:
          message.allowUpdateAfterExpiry = reader.bool();
          break;

        case 11:
          message.allowUpdateAfterMisbehaviour = reader.bool();
          break;

        default:
          reader.skipType(tag & 7);
          break;
      }
    }

    return message;
  },

  fromJSON(object: any): ClientState {
    return {
      chainId: isSet(object.chainId) ? String(object.chainId) : \\"\\",
      trustLevel: isSet(object.trustLevel) ? Fraction.fromJSON(object.trustLevel) : undefined,
      trustingPeriod: isSet(object.trustingPeriod) ? String(object.trustingPeriod) : undefined,
      unbondingPeriod: isSet(object.unbondingPeriod) ? String(object.unbondingPeriod) : undefined,
      maxClockDrift: isSet(object.maxClockDrift) ? String(object.maxClockDrift) : undefined,
      frozenHeight: isSet(object.frozenHeight) ? Height.fromJSON(object.frozenHeight) : undefined,
      latestHeight: isSet(object.latestHeight) ? Height.fromJSON(object.latestHeight) : undefined,
      proofSpecs: Array.isArray(object?.proofSpecs) ? object.proofSpecs.map((e: any) => ProofSpec.fromJSON(e)) : [],
      upgradePath: Array.isArray(object?.upgradePath) ? object.upgradePath.map((e: any) => String(e)) : [],
      allowUpdateAfterExpiry: isSet(object.allowUpdateAfterExpiry) ? Boolean(object.allowUpdateAfterExpiry) : false,
      allowUpdateAfterMisbehaviour: isSet(object.allowUpdateAfterMisbehaviour) ? Boolean(object.allowUpdateAfterMisbehaviour) : false
    };
  },

  toJSON(message: ClientState): unknown {
    const obj: any = {};
    message.chainId !== undefined && (obj.chainId = message.chainId);
    message.trustLevel !== undefined && (obj.trustLevel = message.trustLevel ? Fraction.toJSON(message.trustLevel) : undefined);
    message.trustingPeriod !== undefined && (obj.trustingPeriod = message.trustingPeriod);
    message.unbondingPeriod !== undefined && (obj.unbondingPeriod = message.unbondingPeriod);
    message.maxClockDrift !== undefined && (obj.maxClockDrift = message.maxClockDrift);
    message.frozenHeight !== undefined && (obj.frozenHeight = message.frozenHeight ? Height.toJSON(message.frozenHeight) : undefined);
    message.latestHeight !== undefined && (obj.latestHeight = message.latestHeight ? Height.toJSON(message.latestHeight) : undefined);

    if (message.proofSpecs) {
      obj.proofSpecs = message.proofSpecs.map(e => e ? ProofSpec.toJSON(e) : undefined);
    } else {
      obj.proofSpecs = [];
    }

    if (message.upgradePath) {
      obj.upgradePath = message.upgradePath.map(e => e);
    } else {
      obj.upgradePath = [];
    }

    message.allowUpdateAfterExpiry !== undefined && (obj.allowUpdateAfterExpiry = message.allowUpdateAfterExpiry);
    message.allowUpdateAfterMisbehaviour !== undefined && (obj.allowUpdateAfterMisbehaviour = message.allowUpdateAfterMisbehaviour);
    return obj;
  },

  fromPartial(object: DeepPartial<ClientState>): ClientState {
    const message = createBaseClientState();
    message.chainId = object.chainId ?? \\"\\";
    message.trustLevel = object.trustLevel !== undefined && object.trustLevel !== null ? Fraction.fromPartial(object.trustLevel) : undefined;
    message.trustingPeriod = object.trustingPeriod ?? undefined;
    message.unbondingPeriod = object.unbondingPeriod ?? undefined;
    message.maxClockDrift = object.maxClockDrift ?? undefined;
    message.frozenHeight = object.frozenHeight !== undefined && object.frozenHeight !== null ? Height.fromPartial(object.frozenHeight) : undefined;
    message.latestHeight = object.latestHeight !== undefined && object.latestHeight !== null ? Height.fromPartial(object.latestHeight) : undefined;
    message.proofSpecs = object.proofSpecs?.map(e => ProofSpec.fromPartial(e)) || [];
    message.upgradePath = object.upgradePath?.map(e => e) || [];
    message.allowUpdateAfterExpiry = object.allowUpdateAfterExpiry ?? false;
    message.allowUpdateAfterMisbehaviour = object.allowUpdateAfterMisbehaviour ?? false;
    return message;
  }

};"
`;

exports[`ibc/lightclients/tendermint/v1/tendermint string interface 1`] = `
"/**
 * ClientState from Tendermint tracks the current validator set, latest height,
 * and a possible frozen height.
 */
export interface ClientState {
  chainId: string;
  trustLevel: Fraction;

  /**
   * duration of the period since the LastestTimestamp during which the
   * submitted headers are valid for upgrade
   */
  trustingPeriod: string;

  /** duration of the staking unbonding period */
  unbondingPeriod: string;

  /** defines how much new (untrusted) header's Time can drift into the future. */
  maxClockDrift: string;

  /** Block height when the client was frozen due to a misbehaviour */
  frozenHeight: Height;

  /** Latest height the client was updated to */
  latestHeight: Height;

  /** Proof specifications used in verifying counterparty state */
  proofSpecs: ProofSpec[];

  /**
   * Path at which next upgraded client will be committed.
   * Each element corresponds to the key for a single CommitmentProof in the
   * chained proof. NOTE: ClientState must stored under
   * \`{upgradePath}/{upgradeHeight}/clientState\` ConsensusState must be stored
   * under \`{upgradepath}/{upgradeHeight}/consensusState\` For SDK chains using
   * the default upgrade module, upgrade_path should be []string{\\"upgrade\\",
   * \\"upgradedIBCState\\"}\`
   */
  upgradePath: string[];

  /**
   * This flag, when set to true, will allow governance to recover a client
   * which has expired
   */
  allowUpdateAfterExpiry: boolean;

  /**
   * This flag, when set to true, will allow governance to unfreeze a client
   * whose chain has experienced a misbehaviour event
   */
  allowUpdateAfterMisbehaviour: boolean;
}"
`;

exports[`ibc/lightclients/tendermint/v1/tendermint string proto 1`] = `
"export const ClientState = {
  encode(message: ClientState, writer: _m0.Writer = _m0.Writer.create()): _m0.Writer {
    if (message.chainId !== \\"\\") {
      writer.uint32(10).string(message.chainId);
    }

    if (message.trustLevel !== undefined) {
      Fraction.encode(message.trustLevel, writer.uint32(18).fork()).ldelim();
    }

    if (message.trustingPeriod !== undefined) {
      Duration.encode(message.trustingPeriod, writer.uint32(26).fork()).ldelim();
    }

    if (message.unbondingPeriod !== undefined) {
      Duration.encode(message.unbondingPeriod, writer.uint32(34).fork()).ldelim();
    }

    if (message.maxClockDrift !== undefined) {
      Duration.encode(message.maxClockDrift, writer.uint32(42).fork()).ldelim();
    }

    if (message.frozenHeight !== undefined) {
      Height.encode(message.frozenHeight, writer.uint32(50).fork()).ldelim();
    }

    if (message.latestHeight !== undefined) {
      Height.encode(message.latestHeight, writer.uint32(58).fork()).ldelim();
    }

    for (const v of message.proofSpecs) {
      ProofSpec.encode(v!, writer.uint32(66).fork()).ldelim();
    }

    for (const v of message.upgradePath) {
      writer.uint32(74).string(v!);
    }

    if (message.allowUpdateAfterExpiry === true) {
      writer.uint32(80).bool(message.allowUpdateAfterExpiry);
    }

    if (message.allowUpdateAfterMisbehaviour === true) {
      writer.uint32(88).bool(message.allowUpdateAfterMisbehaviour);
    }

    return writer;
  },

  decode(input: _m0.Reader | Uint8Array, length?: number): ClientState {
    const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseClientState();

    while (reader.pos < end) {
      const tag = reader.uint32();

      switch (tag >>> 3) {
        case 1:
          message.chainId = reader.string();
          break;

        case 2:
          message.trustLevel = Fraction.decode(reader, reader.uint32());
          break;

        case 3:
          message.trustingPeriod = fromDuration(Duration.decode(reader, reader.uint32()));
          break;

        case 4:
          message.unbondingPeriod = fromDuration(Duration.decode(reader, reader.uint32()));
          break;

        case 5:
          message.maxClockDrift = fromDuration(Duration.decode(reader, reader.uint32()));
          break;

        case 6:
          message.frozenHeight = Height.decode(reader, reader.uint32());
          break;

        case 7:
          message.latestHeight = Height.decode(reader, reader.uint32());
          break;

        case 8:
          message.proofSpecs.push(ProofSpec.decode(reader, reader.uint32()));
          break;

        case 9:
          message.upgradePath.push(reader.string());
          break;

        case 10:
          message.allowUpdateAfterExpiry = reader.bool();
          break;

        case 11:
          message.allowUpdateAfterMisbehaviour = reader.bool();
          break;

        default:
          reader.skipType(tag & 7);
          break;
      }
    }

    return message;
  },

  fromJSON(object: any): ClientState {
    return {
      chainId: isSet(object.chainId) ? String(object.chainId) : \\"\\",
      trustLevel: isSet(object.trustLevel) ? Fraction.fromJSON(object.trustLevel) : undefined,
      trustingPeriod: isSet(object.trustingPeriod) ? String(object.trustingPeriod) : undefined,
      unbondingPeriod: isSet(object.unbondingPeriod) ? String(object.unbondingPeriod) : undefined,
      maxClockDrift: isSet(object.maxClockDrift) ? String(object.maxClockDrift) : undefined,
      frozenHeight: isSet(object.frozenHeight) ? Height.fromJSON(object.frozenHeight) : undefined,
      latestHeight: isSet(object.latestHeight) ? Height.fromJSON(object.latestHeight) : undefined,
      proofSpecs: Array.isArray(object?.proofSpecs) ? object.proofSpecs.map((e: any) => ProofSpec.fromJSON(e)) : [],
      upgradePath: Array.isArray(object?.upgradePath) ? object.upgradePath.map((e: any) => String(e)) : [],
      allowUpdateAfterExpiry: isSet(object.allowUpdateAfterExpiry) ? Boolean(object.allowUpdateAfterExpiry) : false,
      allowUpdateAfterMisbehaviour: isSet(object.allowUpdateAfterMisbehaviour) ? Boolean(object.allowUpdateAfterMisbehaviour) : false
    };
  },

  toJSON(message: ClientState): unknown {
    const obj: any = {};
    message.chainId !== undefined && (obj.chainId = message.chainId);
    message.trustLevel !== undefined && (obj.trustLevel = message.trustLevel ? Fraction.toJSON(message.trustLevel) : undefined);
    message.trustingPeriod !== undefined && (obj.trustingPeriod = message.trustingPeriod);
    message.unbondingPeriod !== undefined && (obj.unbondingPeriod = message.unbondingPeriod);
    message.maxClockDrift !== undefined && (obj.maxClockDrift = message.maxClockDrift);
    message.frozenHeight !== undefined && (obj.frozenHeight = message.frozenHeight ? Height.toJSON(message.frozenHeight) : undefined);
    message.latestHeight !== undefined && (obj.latestHeight = message.latestHeight ? Height.toJSON(message.latestHeight) : undefined);

    if (message.proofSpecs) {
      obj.proofSpecs = message.proofSpecs.map(e => e ? ProofSpec.toJSON(e) : undefined);
    } else {
      obj.proofSpecs = [];
    }

    if (message.upgradePath) {
      obj.upgradePath = message.upgradePath.map(e => e);
    } else {
      obj.upgradePath = [];
    }

    message.allowUpdateAfterExpiry !== undefined && (obj.allowUpdateAfterExpiry = message.allowUpdateAfterExpiry);
    message.allowUpdateAfterMisbehaviour !== undefined && (obj.allowUpdateAfterMisbehaviour = message.allowUpdateAfterMisbehaviour);
    return obj;
  },

  fromPartial(object: DeepPartial<ClientState>): ClientState {
    const message = createBaseClientState();
    message.chainId = object.chainId ?? \\"\\";
    message.trustLevel = object.trustLevel !== undefined && object.trustLevel !== null ? Fraction.fromPartial(object.trustLevel) : undefined;
    message.trustingPeriod = object.trustingPeriod ?? undefined;
    message.unbondingPeriod = object.unbondingPeriod ?? undefined;
    message.maxClockDrift = object.maxClockDrift ?? undefined;
    message.frozenHeight = object.frozenHeight !== undefined && object.frozenHeight !== null ? Height.fromPartial(object.frozenHeight) : undefined;
    message.latestHeight = object.latestHeight !== undefined && object.latestHeight !== null ? Height.fromPartial(object.latestHeight) : undefined;
    message.proofSpecs = object.proofSpecs?.map(e => ProofSpec.fromPartial(e)) || [];
    message.upgradePath = object.upgradePath?.map(e => e) || [];
    message.allowUpdateAfterExpiry = object.allowUpdateAfterExpiry ?? false;
    message.allowUpdateAfterMisbehaviour = object.allowUpdateAfterMisbehaviour ?? false;
    return message;
  }

};"
`;
